Fixed the failing specs in rearch and moved presentation out of conductors

parndt opened the pull request:
@robyurkowski please let me know if I screwed something up here.

parndt committed:
Trying to fix up fetches pages.

parndt committed:
Move presentation from conductor to controller.

parndt commented on delivery/backend/app/controllers/abc/backend/pages_controller.rb:
```diff
@@ -20,11 +24,19 @@ def create
+          pages.map { |page| presenters[:pages].new(page) }
+        end
+
+        @data = OpenStruct.new hsh
```
I had to override the way that present was working to add this `map` statement. Please evaluate the existing present because I don't think we should be duplicating if we can help it... and I didn't want to write `if Array === oldval` ;-)

robyurkowski commented on core/abc/conductors/pages/shows_page.rb:
```diff
@@ -26,7 +26,7 @@ def build_page
         private
         def page_repository
-          repositories.page
+          repositories.pages
```
Why not `page`? It's still the page repository even though we're acting on a collection.

robyurkowski commented on delivery/backend/app/conductors/pages/accepts_page_form.rb:
```diff
@@ -8,30 +7,18 @@ module Pages
-            :repositories => { :pages => Adapters::Persistence::Repositories::Page.new },
-            :presenters  => { :pages => Presenters::PagePresenter }
+            :repositories => {
+              :pages => Adapters::Persistence::Repositories::Page.new
```
This is what doesn't make sense to me -- calling this `repositories.pages` when it's the Page repository.

robyurkowski commented on delivery/backend/app/controllers/abc/backend/pages_controller.rb:
```diff
@@ -1,16 +1,20 @@
-require 'pages/accepts_page_form'
-require 'presenters/abc/page_presenter'
+require 'conductors/pages/accepts_page_form'
+require 'abc/presentation'
```
:heart: :+1:

robyurkowski commented on delivery/backend/app/controllers/abc/backend/pages_controller.rb:
```diff
@@ -1,16 +1,20 @@
 
       def new
+        # Maybe?
+        # @page = Entities::Page.new
```
I think this still defers to a Conductor, even in simple cases (because there's still a very good chance we could want to fetch menus, for instance).

parndt replied:
What would the conductor be called?

robyurkowski commented on delivery/backend/app/controllers/abc/backend/pages_controller.rb:
```diff
@@ -20,11 +24,19 @@ def create
+
+      def present(data)
+        hsh = data.merge(data) do |key, pages, _|
+          pages.map { |page| presenters[:pages].new(page) }
```
Definitely singular on the presenters, IMO.

parndt committed:
Singular for singular; plural for collections.

robyurkowski commented on delivery/backend/app/views/abc/backend/pages/index.html.haml:
```diff
@@ -1,4 +1,4 @@
 %h2 Pages
 - @data.pages.each do |page|
-  = link_to page, '#'
+  = link_to page.title, '#'
```
Yeah, using `to_s` felt weird here. :+1:.

parndt replied:
It also didn't work anymore after moving to the html presenter.

robyurkowski commented on delivery/backend/spec/integration/pages_spec.rb:
```diff
@@ -43,7 +43,7 @@
 end
 
-def create_page(attributes={})
+def create_page(attributes = {})
```
:clap:

robyurkowski commented on delivery/shared/lib/abc/presentation.rb:
```diff
@@ -0,0 +1,33 @@
+    # can be accessed in the view.
+    #
+    # @param hsh [Hash] Data to be presented to the view
+    def present(data)
```
How about...

``` ruby
def present(data)
  hsh = data.merge(data) do |key, oldval, _|
    if presenters.has_key?(key)
      presenters[key].new(oldval)
    elsif oldval.is_a?(Array) && presenters.has_key?(key.singularize)
      oldval.map {|item| presenters[key.singularize].new(item) }
    else
      oldval
    end
  end

  @data = OpenStruct.new hsh
end
```

parndt replied:
What's the purpose of the singularize in the `elsif`? 
I'm super unhappy about having such an ugly conditional but I'm not sure what can be done about that right now. Perhaps we can refactor it later.

parndt replied:
I've implemented this and now `abc-delivery-shared` depends on `activesupport` to use `active_support/core_ext/string` so that `singularize` exists.

parndt committed:
Made the Abc::Presentation#present method more flexible.

parndt committed:
Moved create_pages migration to delivery/shared

parndt committed:
Made abc-delivery-shared a Rails Engine.

parndt committed:
Load the shared engine when Rails is around.

Also improve test coverage by using the public interface of FetchesPages instead of send(:data)

knewter commented on delivery/backend/app/controllers/abc/backend/pages_controller.rb:
```diff
@@ -1,16 +1,20 @@
+
       def index
-        present Conductors::Pages::FetchesPages.new(params, :repositories => repositories)
+        present Conductors::Pages::FetchesPages.new(params, :repositories => repositories).call
```
I like the way this pattern looks more and more now (`#present` is great imo), and with a factory of some sort to handle the di bits (like how obvious does it, other repository pattern projects I've seen) it would be un-verbose as well.  Could eventually look like:

``` ruby
present Conductors::Pages::FetchesPage # in the simple case
```

I'm not advocating for this yet, and I don't think that's the end-all-be-all of the potential sugar for this, but I like that this pattern seems eminently shorthandable

robyurkowski replied:
I think with a decent DSL, this'll be very useful.
